File: plugin-tester-java/src/main/java/io/grpc/examples/helloworld/GreeterServiceImpl.java
//#full-service-impl
package io.grpc.examples.helloworld;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.stream.Collectors;

import akka.NotUsed;
import akka.stream.Materializer;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;

public class GreeterServiceImpl implements GreeterService {
  private final Materializer mat;

  public GreeterServiceImpl(Materializer mat) {
   this.mat = mat;
  }


  @Override
  public CompletionStage<HelloReply> sayHello(HelloRequest in) {
    System.out.println("sayHello to " + in.getName());
    HelloReply reply = HelloReply.newBuilder().setMessage("Hello, " + in.getName()).build();
    return CompletableFuture.completedFuture(reply);
  }

  @Override
  public CompletionStage<HelloReply> itKeepsTalking(Source<HelloRequest, Object> in) {
    System.out.println("sayHello to in stream...");
    return in.runWith(Sink.seq(), mat)
      .thenApply(elements -> {
        return HelloReply.newBuilder().setMessage("Hello, " + elements).build();
      });
  }

  @Override
  public Source<HelloReply, Object> itKeepsReplying(HelloRequest in) {
    System.out.println("sayHello to " + in.getName() + " with stream of chars");
    List<Character> characters = ("Hello, " + in.getName())
        .chars().mapToObj(c -> (char) c).collect(Collectors.toList());
    return Source.from(characters)
      .map(character -> {
        return HelloReply.newBuilder().setMessage(String.valueOf(character)).build();
      })
      .mapMaterializedValue(m -> NotUsed.getInstance()); // FIXME would be nice without this
  }

  @Override
  public Source<HelloReply, Object> streamHellos(Source<HelloRequest, Object> in) {
    System.out.println("sayHello to stream...");
    return in.map(request -> HelloReply.newBuilder().setMessage("Hello, " + request.getName()).build());
  }
}
//#full-service-impl

File: plugin-tester-scala/src/main/scala/example/myapp/helloworld/GreeterServiceImpl.scala
/*
 * Copyright (C) 2018-2021 Lightbend Inc. <https://www.lightbend.com>
 */

//#full-service-impl
package example.myapp.helloworld

import scala.concurrent.Future
import akka.NotUsed
import akka.stream.Materializer
import akka.stream.scaladsl.Sink
import akka.stream.scaladsl.Source
import com.google.protobuf.timestamp.Timestamp
import example.myapp.helloworld.grpc.GreeterServiceFilterApi.GenericFilter
import example.myapp.helloworld.grpc._
import scala.concurrent.duration._
class GreeterServiceImpl(implicit mat: Materializer) extends GreeterService {
  import mat.executionContext

  override def sayHello(in: HelloRequest): Future[HelloReply] = {
    println(s"sayHello to ${in.name}")
    lazy val response = Future.successful(HelloReply(s"Hello, ${in.name}", Some(Timestamp.apply(123456, 123))))
    if (in.name.contains("sleep")) {
      akka.pattern.after(5.second, mat.system.scheduler)(response)
    } else response
  }

  override def itKeepsTalking(in: Source[HelloRequest, NotUsed]): Future[HelloReply] = {
    println(s"sayHello to in stream...")
    in.runWith(Sink.seq)
      .map(elements => HelloReply(s"Hello, ${elements.map(_.name).mkString(", ")}", Some(Timestamp.apply(123456, 123))))
  }

  override def itKeepsReplying(in: HelloRequest): Source[HelloReply, NotUsed] = {
    println(s"sayHello to ${in.name} with stream of chars...")
    Source(s"Hello, ${in.name}".toList).map(character => HelloReply(character.toString))
  }

  override def streamHellos(in: Source[HelloRequest, NotUsed]): Source[HelloReply, NotUsed] = {
    println(s"sayHello to stream...")
    in.map(request => HelloReply(s"Hello, ${request.name}", Some(Timestamp.apply(123456, 123))))
  }

  val FilterItems = new FilterItems(mat.system)

  override val sayHelloFilter: Seq[GenericFilter[HelloRequest, Future[HelloReply]]] =
    Seq(
      FilterItems.RequestResponse.requestDuration,
      FilterItems.RequestResponse.timeout,
      FilterItems.RequestResponse.add,
      FilterItems.RequestResponse.remove)

  override val itKeepsTalkingFilter
      : Seq[GreeterServiceFilterApi.GenericFilter[Source[HelloRequest, NotUsed], Future[HelloReply]]] =
    Seq(FilterItems.StreamRequestResponse.add, FilterItems.StreamRequestResponse.remove)

  override val itKeepsReplyingFilter
      : Seq[GreeterServiceFilterApi.GenericFilter[HelloRequest, Source[HelloReply, akka.NotUsed]]] =
    Seq(FilterItems.RequestStreamResponse.add, FilterItems.RequestStreamResponse.remove)

  override val streamHellosFilter
      : Seq[GreeterServiceFilterApi.GenericFilter[Source[HelloRequest, NotUsed], Source[HelloReply, akka.NotUsed]]] =
    Seq(FilterItems.StreamRequestStreamResponse.add, FilterItems.StreamRequestStreamResponse.remove)

}
//#full-service-impl
